File: csharp/lang/scheduling/in-process-scheduler.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Hangfire;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;

namespace Example.Orders;

public class Rollups
{
    private Timer? timer;

    public void StartReconciler()
    {
        // ruleid: in-process-scheduler
        timer = new Timer(_ => Reconcile(), null, TimeSpan.Zero, TimeSpan.FromMinutes(5));
    }

    public void ConfigureHangfire(IServiceCollection services)
    {
        // ruleid: in-process-scheduler
        services.AddHangfire(config => config.UseMemoryStorage());

        // ruleid: in-process-scheduler
        RecurringJob.AddOrUpdate("nightly", () => Reconcile(), Cron.Daily);
    }

    public void ConfigureShared(IServiceCollection services, string connectionString)
    {
        // ok: in-process-scheduler
        services.AddHangfire(config => config.UsePostgreSqlStorage(connectionString));
    }

    private void Reconcile() { }
}

// ok: in-process-scheduler
public class TriggeredWorker : BackgroundService
{
    protected override Task ExecuteAsync(CancellationToken stoppingToken) => Task.CompletedTask;
}

File: python/lang/scheduling/in-process-scheduler.py
import threading

import schedule
from apscheduler.schedulers.background import BackgroundScheduler
from celery import Celery


def start_reconciler(reconcile):
    # ruleid: in-process-scheduler
    scheduler = BackgroundScheduler()
    scheduler.add_job(reconcile, "interval", minutes=5)
    scheduler.start()


def register_nightly(rollup):
    # ruleid: in-process-scheduler
    schedule.every().day.at("02:00").do(rollup)


def delayed_retry(fn):
    # ruleid: in-process-scheduler
    threading.Timer(30.0, fn).start()


app = Celery("tasks")

# ok: in-process-scheduler
app.conf.beat_schedule = {
    "nightly-rollup": {"task": "tasks.rollup", "schedule": 86400.0},
}


# ok: in-process-scheduler
@app.task
def rollup():
    return True
